File: openc3/lib/openc3/accessors/xml_accessor.rb
# encoding: ascii-8bit

# Copyright 2022 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'nokogiri'
require 'openc3/accessors/accessor'

module OpenC3
  class XmlAccessor < Accessor
    def self.read_item(item, buffer)
      return nil if item.data_type == :DERIVED
      doc = buffer_to_doc(buffer)
      return convert_to_type(doc.xpath(item.key).first.to_s, item)
    end

    def self.write_item(item, value, buffer)
      return nil if item.data_type == :DERIVED
      doc = buffer_to_doc(buffer)
      node = doc.xpath(item.key).first
      node.content = value.to_s
      buffer.replace(doc_to_buffer(doc))
      return value
    end

    def self.read_items(items, buffer)
      doc = buffer_to_doc(buffer)
      result = {}
      items.each do |item|
        if item.data_type == :DERIVED
          result[item.name] = nil
        else
          result[item.name] = convert_to_type(doc.xpath(item.key).first.to_s, item)
        end
      end
      return result
    end

    def self.write_items(items, values, buffer)
      doc = buffer_to_doc(buffer)
      items.each_with_index do |item, index|
        next if item.data_type == :DERIVED
        node = doc.xpath(item.key).first
        node.content = values[index].to_s
      end
      buffer.replace(doc_to_buffer(doc))
      return values
    end

    def self.buffer_to_doc(buffer)
      Nokogiri.XML(buffer)
    end

    def self.doc_to_buffer(doc)
      doc.to_xml
    end

    # If this is set it will enforce that buffer data is encoded
    # in a specific encoding
    def enforce_encoding
      return nil
    end

    # This affects whether the Packet class enforces the buffer
    # length at all.  Set to false to remove any correlation between
    # buffer length and defined sizes of items in COSMOS
    def enforce_length
      return false
    end

    # This sets the short_buffer_allowed flag in the Packet class
    # which allows packets that have a buffer shorter than the defined size.
    # Note that the buffer is still resized to the defined length
    def enforce_short_buffer_allowed
      return true
    end

    # If this is true it will enfore that COSMOS DERIVED items must have a
    # write_conversion to be written
    def enforce_derived_write_conversion(_item)
      return true
    end
  end
end

File: openc3/python/openc3/accessors/xml_accessor.py
# Copyright 2024 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

from .accessor import Accessor
# HtmlAccessor uses lxml.html
import lxml.etree as ET


class XmlAccessor(Accessor):
    @classmethod
    def class_read_item(cls, item, buffer):
        if item.data_type == "DERIVED":
            return None
        doc = cls._buffer_to_doc(buffer)
        doc_value = doc.xpath(item.key)
        # The xpath method returns a boolean, float, string, or list of items
        # If it's a list then just grab the first value to convert
        if isinstance(doc_value, list):
            doc_value = doc_value[0]
        return cls.convert_to_type(doc_value, item)

    @classmethod
    def class_write_item(cls, item, value, buffer):
        if item.data_type == "DERIVED":
            return None
        doc = cls._buffer_to_doc(buffer)
        cls._update_doc(doc, item, value)
        buffer[0:] = cls._doc_to_buffer(doc)

    @classmethod
    def class_read_items(cls, items, buffer):
        doc = cls._buffer_to_doc(buffer)
        result = {}
        for item in items:
            if item.data_type == "DERIVED":
                result[item.name] = None
            else:
                doc_value = doc.xpath(item.key)
                # The xpath method returns a boolean, float, string, or list of items
                # If it's a list then just grab the first value to convert
                if isinstance(doc_value, list):
                    doc_value = doc_value[0]
                result[item.name] = cls.convert_to_type(doc_value, item)
        return result

    @classmethod
    def class_write_items(cls, items, values, buffer):
        doc = cls._buffer_to_doc(buffer)
        for index, item in enumerate(items):
            cls._update_doc(doc, item, values[index])
        buffer[0:] = cls._doc_to_buffer(doc)

    @classmethod
    def _buffer_to_doc(cls, buffer):
        return ET.fromstring(buffer.decode())

    @classmethod
    def _doc_to_buffer(cls, doc):
        return bytearray(ET.tostring(doc))

    @classmethod
    def _update_doc(cls, doc, item, value):
        # Split the attribute or text selector from the xpath so we get a
        # lxml.etree._Element that we can operate on. Otherwise we get a
        # lxml.etree._ElementUnicodeResult which is read-only.
        parts = item.key.split("/")
        path = "/".join(parts[0:-1])
        node = doc.xpath(path)[0]
        # Determine what the selector was trying to set
        if "@" in parts[-1]:
            node.attrib[parts[-1][1:]] = str(value)
        elif "text()" == parts[-1]:
            node.text = str(value)
        else:
            raise RuntimeError(f"Unknown selector: {item.key}")

    def enforce_encoding(self):
        return None

    def enforce_length(self):
        return False

    def enforce_short_buffer_allowed(self):
        return True

    def enforce_derived_write_conversion(self, item):
        return True
